### PR TITLE
Track the entire scanned range in `DbIteratorRangeTracker` to fix some phantom reads and write skew

### LINKED ISSUE #1711: SSI range tracking only records yielded keys, allowing phantom reads / write skew on `scan` and `scan_prefix`
**Describe the bug**

I think the `DbIteratorRangeTracker` is accidentally tracking only entries that were read, instead of the entire range, this can cause phantom reads and write skew in some situations. It records only the rows actually yielded by the iterator, not the range the caller requested. This means that it won't properly track a range without data, allowing for write skew. Or it could also miss tombstones, or have data come back in a transaction with `IsolationLevel::SerializableSnapshot` set. It accidentally does the naive implementation described here: https://github.com/slatedb/slatedb/blob/main/slatedb/src/db_iter.rs#L25-L28

As a result:

- An empty scan registers no range at all (`has_data()` stays false), so any concurrent write inside the requested range is undetected.
- A partial scan registers only `[first_yielded_key, last_yielded_key]`, so a concurrent insert into the unobserved portion of the requested range is also undetected.

Because `transaction_manager.rs::has_read_write_conflict` only consults the yielded rows, invariants of serializable snapshot isolation level are broken.

**Root cause (summary)**

- `DbIteratorRangeTracker::track_key` only updates `first_key`/`last_key`, and `get_range()` returns `None` if nothing was yielded.
- `track_key` is only invoked from `DbIterator::next_entry` for entries actually returned to the caller. Tombstones and unread suffixes never contribute, and the requested range bounds passed to `DbIterator::new` are never given to the tracker.
- on commit, if `has_data()` is false the range is silently dropped, so the rw-conflict check has nothing to compare against.

**To Reproduce**

See regression tests: https://github.com/slatedb/slatedb/pull/1710/changes/f211cc478e5f2322454b9636bca75cf89baf3db9

Add `slatedb/tests/ssi_phantom_repro.rs` with three reproducers and run:

```
cargo test --package slatedb --test ssi_phantom_repro -- --nocapture
```

All three fail on `main` @ 0864ef6:

```
test ssi_empty_scan_prefix_misses_concurrent_insert        ... FAILED
test ssi_partial_scan_misses_phantom_outside_yielded_bbox  ... FAILED
test ssi_write_skew_via_empty_scan_phantom                 ... FAILED
```

**Test 1, write-skew anomaly**

[impl here](https://github.com/slatedb/slatedb/pull/1710/changes/f211cc478e5f2322454b9636bca75cf89baf3db9#diff-2d2522258ce0ed49a70728a5b0589a2a5ade19be8e5cd84d9a71691c53dbd494R958)

```
T1 = db.begin(SerializableSnapshot)
T2 = db.begin(SerializableSnapshot)
T1.scan("a".."m")   -> empty       // requested range tracked = NONE
T2.scan("n".."z")   -> empty       // requested range tracked = NONE
T1.put("x_from_t1", ...)           // x is in T2's scanned range
T2.put("b_from_t2", ...)           // b is in T1's scanned range
T1.commit() -> Ok                  // EXPECTED: exactly one conflicts
T2.commit() -> Ok                  // OBSERVED: BOTH commit
```

Neither serial order is consistent with the observed reads, which means it is non-serializable.

**Test 2 Phantom outside the yielded bounding box**

[impl here](https://github.com/slatedb/slatedb/pull/1710/changes/f211cc478e5f2322454b9636bca75cf89baf3db9#diff-2d2522258ce0ed49a70728a5b0589a2a5ade19be8e5cd84d9a71691c53dbd494R1006)

```
db.put("k_mid", "seed")
T1 = db.begin(SerializableSnapshot)
iter = T1.scan("k_aaa".."k_zzz")    // yields only k_mid; tracker = [k_mid, k_mid]
db.put("k_aaa_inserted", "phantom") // INSIDE T1's scan range, outside box
T1.put("t1_marker", "x")
T1.commit() -> Ok                   // OBSERVED: commits silently
```

Expected `TransactionConflict`.

**Test 3:  Empty prefix-scan blind-spot ("first user" gate)**

[impl here](https://github.com/slatedb/slatedb/pull/1710/changes/f211cc478e5f2322454b9636bca75cf89baf3db9#diff-2d2522258ce0ed49a70728a5b0589a2a5ade19be8e5cd84d9a71691c53dbd494R1078)

```
T1 = db.begin(SerializableSnapshot)
T1.scan_prefix("users/") -> empty   // tracker = NONE
db.put("users/alice", "first")      // concurrent writer
T1.put("users/bob", "second")
T1.commit() -> Ok                   // OBSERVED: commits silently
```

Final state contains both `users/alice` and `users/bob` despite T1 gating on an empty `users/` prefix. The "uniqueness gate" pattern is silently broken.

**Additional tombstone resurrection case**: https://github.com/slatedb/slatedb/pull/1710/changes/f211cc478e5f2322454b9636bca75cf89baf3db9#diff-2d2522258ce0ed49a70728a5b0589a2a5ade19be8e5cd84d9a71691c53dbd494R1036-R1070

**Expected behavior**

SSI should track the **requested** scan range, not the yielded bounding box, so that any concurrent write inside the requested range — including inserts into an empty or sparsely-populated range — is detected as a read-write conflict at commit time. This is the guarantee promised by:

- `rfcs/0011-transaction.md:125`: "use Serializable Snapshot Isolation (SSI) level … thus no longer allow the write skew anomaly to occur."
- `slatedb/src/db_iter.rs:21-28`: doc comment on `DbIteratorRangeTracker`, which explicitly describes the phantom-read scenario this implementation fails to prevent.
- `slatedb/src/transaction_manager.rs:13-22`: public `IsolationLevel::SerializableSnapshot` doc: "detects both read-write and write-write conflicts."

**Suggested fix**

Track the requested scan range, not the yielded box:

- In `db_transaction.rs::scan_inner` (and the prefix path), register the user-supplied `BytesRange` directly with `txn_manager.track_read_range` at scan-start under SSI, **or**
- Initialise `DbIteratorRangeTracker` from the requested range and only narrow it when the user terminates the iterator early via `seek`.

The existing `has_read_write_conflict` check in `transaction_manager.rs:346-390` is then sufficient — no algorithmic change needed. Also audit `scan_prefix_with_options` so the registered range corresponds to `BytesRange::from_prefix(...)`, and add SSI tests covering empty scans, partial scans terminated by `seek`, and prefix scans against a concurrent first-writer.

Note: existing tests (`db_transaction.rs:906-947` `test_txn_ssi_commit_conflit_with_ranges`, `transaction_manager.rs:1068-1090` `test_ssi_phantom_read_conflict_on_range`) only exercise the path where the iterator yielded the conflicting key. Neither covers the empty-scan or outside-yielded-box case.

**Screenshots**

N/A

**Build**

`uname -a`: `Linux nixos 6.18.23 #1-NixOS SMP PREEMPT_DYNAMIC Sat Apr 18 08:45:02 UTC 2026 x86_64 GNU/Linux`

- OS: NixOS (Linux 6.18.23)
- SlateDB Version: 0.13.0 (`main` @ 0864ef6)
- Object Store: in-memory (reproducer uses default test object store)

**Additional context**

- Severity: High — silent serializability violation in the documented strict isolation level.
- Scope: limited to SSI range-tracking. SI guarantees are unaffected (SI never promised to prevent write skew). Write-write conflict detection, point-read tracking, read-your-own-writes, snapshot visibility, and the single-writer commit pipeline were all inspected and appear correct on `main`.

### PR DESCRIPTION
**Pull request by kirinrastogi**
## Summary

In serializable snapshot isolation mode, ranges of scans need to be tracked so that conflicts in the ranges can be detected, and transactions can be aborted. However, with the current implementation, only yielded rows are checked, and not the entire scanned range.

This means that ranges without data, or ranges containing tombstones may be vulnerable to write skew or phantom reads - as the range tracker won't consider those entries.

The easiest way to conceptualize this will also go in a bug report, but in `test_txn_ssi_write_skew_via_empty_scans` we see that two transactions can write to each other's ranges (which are empty) and both can succeed. But that means that a read showed up later in one of the transactions and it was allowed to commit.

Another risk is for tombstones, as tombstones aren't yielded to the range tracker. In `test_txn_ssi_tombstone_resurrection_in_scan_range` we see that a tombstone can be resurrected during a transaction, and it is allowed to commit.

Fixes https://github.com/slatedb/slatedb/issues/1711

## Changes

To fix this, I think the `DbIteratorRangeTracker` should be simplified to track the requested range, and not the yielded rows. Since it is being dropped from other callers, I also dropped the `Arc<>`.

## Notes for Reviewers

Any hints on how to best review this PR? Anything you’d like reviewers to focus on? Any follow-ups planned?

## Checklist

- [x] Small, scoped PR (< 500 total lines excluding tests); or opened as Draft with a plan on how to break it into smaller pieces
- [x] Linked related issue(s) or added context in the description
- [x] Self-reviewed the diff; added comments for tricky parts
- [x] Tests added/updated and passing locally
- [x] Ran `cargo fmt`, `cargo clippy --all-targets --all-features`, and `cargo nextest run --all-features`
- [x] Called out any breaking changes and provided migration notes
- [x] Considered performance impact; added notes or benchmarks if relevant

Thank you for the review! 🙏

### CODE REVIEW

File: slatedb/src/db_iter.rs [hunk 1]
```diff
@@ -13,82 +13,35 @@ use crate::types::{KeyValue, RowEntry, ValueDeletable};
...
 pub(crate) struct DbIteratorRangeTracker {
-    inner: Mutex<DbIteratorRangeTrackerInner>,
-}
-
-#[derive(Debug)]
-struct DbIteratorRangeTrackerInner {
-    first_key: Option<Bytes>,
-    last_key: Option<Bytes>,
-    has_data: bool,
+    requested_range: BytesRange,
```
Comment (kirinrastogi) [inline slatedb/src/db_iter.rs:35]: Maybe this type is now redundant? I felt like keeping it as a note that db_iter and db_transaction are related.

**Review by rodesai — COMMENTED** [review]
This patch makes sense to me, and seems like the right way to detect SSI conflicts in the presence of range scans. I'd like @flaneur2020 to have a look though - there may be some reason I'm not grokking for using the observed range rather than the requested range.

**Review by criccomini — APPROVED** [review]
LGTM!

**Comment by flaneur2020** [comment]
sorry for the late response, and thanks for the fix!

i previously worried that the actual range scanned by the KV engine might be smaller than the requested scan range. i thought that by covering the requested range, we could cover all gaps within it, which i believed would be equivalent to a gap lock.

looking back, that understanding was incorrect >.<. two empty scan requests followed by an insert is indeed a classic write skew scenario. you're right, basing it on the requested range is the more appropriate approach!
